### PR TITLE
Support for identifying devices by id such as mac address

### PR DESCRIPTION
**Pull request by neilcook**
### Short description
This PR provides support for a deviceId field, set by Lua, and included in protobuf messages. I essentially just shadowed what was already done for requestorID.

DeviceId would typically be parsed from an EDNS option such as 65001 or the new edns0-clientid draft, and would usually be a mac address but could also be an IP 4/6 address.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code and tested it
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Comment by pieterlexis** [comment]
Please add some words about the change in gettag [here](https://doc.powerdns.com/recursor/lua-scripting/hooks.html#gettag) in [this file](https://github.com/PowerDNS/pdns/blob/master/pdns/recursordist/docs/lua-scripting/hooks.rst). I think it can go in the same block that begins with "It can also return a table whose keys and values are strings to fill the ...."

**Comment by Habbie** [comment]
Please add support to https://github.com/PowerDNS/pdns/blob/master/contrib/ProtobufLogger.py as well - looks like requestorId is missing there as well.

**Comment by neilcook** [comment]
Also I need to add documentation at https://github.com/PowerDNS/pdns/blob/master/pdns/recursordist/docs/lua-scripting/dq.rst

File: contrib/ProtobufLogger.py [hunk 1]
```diff
@@ -144,16 +144,20 @@ def printSummary(self, msg, typestr):
...
-                                                                           typestr,
-                                                                           msg.inBytes,
-                                                                           ipfromstr,
-                                                                           requestorstr,
-                                                                           iptostr,
-                                                                           protostr,
-                                                                           msg.id,
-                                                                           messageidstr,
-                                                                           initialrequestidstr))
+        deviceId = binascii.hexlify(bytearray(msg.deviceId))
```
Comment (pieterlexis) [inline contrib/ProtobufLogger.py]: What happens if `msg.deviceId` is empty?

**Comment by Habbie** [comment]
```
Exception in thread Connection Handler:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "ProtobufLogger.py", line 30, in run
    self.printQueryMessage(msg)
  File "ProtobufLogger.py", line 43, in printQueryMessage
    self.printSummary(message, 'Query')
  File "ProtobufLogger.py", line 160, in printSummary
    initialrequestidstr))
TypeError: not enough arguments for format string
```

on the very first log message. This is without deviceid in the message.

File: contrib/ProtobufLogger.py [hunk 2]
```diff
@@ -144,16 +144,20 @@ def printSummary(self, msg, typestr):
...
+                  'requestorid: %s deviceid: %s' % (datestr,
+                                                    typestr,
+                                                    msg.inBytes,
+                                                    ipfromstr,
+                                                    requestorstr,
+                                                    iptostr,
+                                                    protostr,
+                                                    msg.id,
+                                                    messageidstr,
+                                                    initialrequestidstr))
```
Comment (Habbie) [inline contrib/ProtobufLogger.py]: requestorid and deviceid are missing in this parameter list

Comment (Habbie) [reply contrib/ProtobufLogger.py]: `[2017-08-28 16:46:49.461187] Response of size 50: 127.0.0.1 -> 127.0.0.1 (UDP), id: 47036, uuid: f2b4bcba0b144a09b9a49b234770f142 requestorid: pietje deviceid: 414243444546`

With:
```diff
-                                                    initialrequestidstr))
+                                                    initialrequestidstr, requestorId, deviceId))
```

**Comment by neilcook** [comment]
These changes are now tested:

Without deviceId:
[2017-08-29 11:16:45.855169] Query of size 51: 127.0.0.1 -> 127.0.0.1 (UDP), id: 35564, uuid: 407f62d12a484a3191ce3ded52b85fdf requestorid:  deviceid: 
[2017-08-29 11:16:45.855198] Response of size 55: 127.0.0.1 -> 127.0.0.1 (UDP), id: 35564, uuid: 407f62d12a484a3191ce3ded52b85fdf requestorid:  deviceid: 

With deviceId (and requestorId):
[2017-08-29 10:59:25.878745] Query of size 51: 127.0.0.1 -> 127.0.0.1 (UDP), id: 12963, uuid: 55c71a178a2d4f49a1cbecb5756dac5a requestorid: pietje deviceid: 414243444546
[2017-08-29 10:59:25.878774] Response of size 55: 127.0.0.1 -> 127.0.0.1 (UDP), id: 12963, uuid: 55c71a178a2d4f49a1cbecb5756dac5a requestorid: pietje deviceid: 414243444546

File: contrib/ProtobufLogger.py (hunk 1 above)
Comment (neilcook) [reply contrib/ProtobufLogger.py]: python protobuf returns empty python string

**Comment by neilcook** [comment]
JFYI I don't have permission to merge to pdns

**Pull request merged**